shared-bindings/analogio/AnalogFastIn.c

```diff
@@ -0,0 +1,197 @@
+#include "py/binary.h"
+#include "py/mphal.h"
+#include "py/nlr.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared-bindings/analogio/AnalogFastIn.h"
+#include "shared-bindings/util.h"
+
+
+
+//| class AnalogFastIn:
+//|     """Read analog voltage levels quickly using DMA Capture"""
+//|
+//|     def __init__(self, pin: microcontroller.Pin, buffer: ReadableBuffer, *, sample_rate: int = 500000) -> None:
+//|         """Use the AnalogFastIn on the given pin. Fill the given buffer from ADC read values at the supplied
+///         sample_rate.
+///
+//|         :param ~microcontroller.Pin pin: the pin to read from"""
+//|         :param ~circuitpython_typing.WriteableBuffer buffer: A buffer for samples
+//|         :param int sample_rate: The desired playback sample rate
+///
+//|     Usage::
+///
+//|       import board
+//|       import analogio
+//|       import array
+///
+//|       length = 1000
+//|       mybuffer = array.array("H", [0] * length)
+//|       fadc = analogio.AnalogFastIn(board.GP26, mybuffer)
+//|       fadc.capture()
+//|       fadc.deinit()
+//|       for i in range(length):
+//|           print(i, mybuffer[i])
+///
+///       (Future) The reference voltage varies by platform so use ``reference_voltage`` to read the configured setting.
+///       """
+//|         ...
+///
+STATIC void validate_rate(mp_float_t rate) {
+    if (rate < (mp_float_t)1.0f || rate > (mp_float_t)500000.0f) {
+        mp_raise_ValueError(translate("sample rate must be 1.0-500000.0 per second"));
+    }
+}
+
+STATIC mp_obj_t analogio_analogfastin_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_pin, ARG_buffer, ARG_sample_rate };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_pin,    MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_buffer, MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_sample_rate, MP_ARG_OBJ | MP_ARG_REQUIRED, {.u_obj = MP_OBJ_NEW_SMALL_INT(1)} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Validate Pin
+    const mcu_pin_obj_t *pin = validate_obj_is_free_pin(args[ARG_pin].u_obj);
+
+    // Buffer Pointer defined and allocated by user
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(args[ARG_buffer].u_obj, &bufinfo, MP_BUFFER_READ);
+
+    // signed or unsigned, byte per sample
+    bool signed_samples = bufinfo.typecode == 'b' || bufinfo.typecode == 'h';
+    uint8_t bytes_per_sample = 1;
+
+    // Bytes Per Sample
+    if (bufinfo.typecode == 'h' || bufinfo.typecode == 'H') {
+        bytes_per_sample = 2;
+    } else if (bufinfo.typecode != 'b' && bufinfo.typecode != 'B' && bufinfo.typecode != BYTEARRAY_TYPECODE) {
+        mp_raise_ValueError(translate("sample_source buffer must be a bytearray or array of type 'h', 'H', 'b' or 'B'"));
+    }
```

Comment:
Look around for an existing message you can reuse for this, if possible.

Reply:
This ValueError msg was used in audiocore/RawSample.c. How would one establish re-use?